src/tests/UnitTests/UI/TreeBaseTests.cs

```diff
@@ -606,6 +611,196 @@ public void ShouldPopulateTreeWithSingleEntryWithMetaInPath()
+                    Path = "Root\\Parent\\A.txt",
+                    Label = "A.txt",
+                    Level = 3,
+                    TreeData = testTreeData[0],
+                },
+                new TestTreeNode {
+                    Path = "Root\\Parent\\B.txt",
+                    Label = "B.txt",
+                    Level = 3,
+                    TreeData = testTreeData[1],
+                },
+                new TestTreeNode {
+                    Path = "Root\\Parent\\C.txt",
+                    Label = "C.txt",
+                    Level = 3,
+                    TreeData = testTreeData[2],
+                }
+            });
+
+            var rootNode = testTree.CreatedTreeNodes[1];
+            var parentNode = testTree.CreatedTreeNodes[2];
+            var aNode = testTree.CreatedTreeNodes[3];
+            var bNode = testTree.CreatedTreeNodes[4];
+            var cNode = testTree.CreatedTreeNodes[5];
+
+            // Initial state, everything unchecked
+
+            Assert.AreEqual(CheckState.Empty, rootNode.CheckState);
+            Assert.AreEqual(CheckState.Empty, parentNode.CheckState);
+            Assert.AreEqual(CheckState.Empty, aNode.CheckState);
+            Assert.AreEqual(CheckState.Empty, bNode.CheckState);
+            Assert.AreEqual(CheckState.Empty, cNode.CheckState);
+
+            testTree.ToggleNodeChecked(1, rootNode);
+
+            // Checked the root node, everything checked
+
+            Assert.AreEqual(CheckState.Checked, rootNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, parentNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, aNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, bNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, cNode.CheckState);
+
+            testTreeListener.Received(3).AddCheckedNode(Arg.Any<TestTreeNode>());
+            testTreeListener.ClearReceivedCalls();
+
+            // Unchecked c.txt, c.txt unchecked, parents mixed
+
+            testTree.ToggleNodeChecked(5, cNode);
+
+            Assert.AreEqual(CheckState.Mixed, rootNode.CheckState);
+            Assert.AreEqual(CheckState.Mixed, parentNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, aNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, bNode.CheckState);
+            Assert.AreEqual(CheckState.Empty, cNode.CheckState);
+
+            testTreeListener.Received(1).RemoveCheckedNode(Arg.Any<TestTreeNode>());
+            testTreeListener.ClearReceivedCalls();
+
+            testTree.ToggleNodeChecked(5, cNode);
+
+            // Checked c.txt, everything checked
+
+            Assert.AreEqual(CheckState.Checked, rootNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, parentNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, aNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, bNode.CheckState);
+            Assert.AreEqual(CheckState.Checked, cNode.CheckState);
+
```

Comment:
Can we do one more where we uncheck `a.txt`, `b.txt`, and `c.txt`, and verify that the parent unchecks entirely?

Reply:
Actually, the whole tree should uncheck at that point, no? That feels like a really appropriate final assertion. ;)

Reply:
No problem